File: app/src/main/java/me/devsaki/hentoid/fragments/reader/ReaderPrefsDialogFragment.java
package me.devsaki.hentoid.fragments.reader;

import static me.devsaki.hentoid.util.Preferences.Key.VIEWER_BROWSE_MODE;
import static me.devsaki.hentoid.util.Preferences.Key.VIEWER_IMAGE_DISPLAY;
import static me.devsaki.hentoid.util.Preferences.Key.VIEWER_RENDERING;

import android.content.Intent;
import android.content.res.Resources;
import android.os.Build;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.DialogFragment;
import androidx.fragment.app.Fragment;

import com.skydoves.powerspinner.PowerSpinnerView;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import me.devsaki.hentoid.R;
import me.devsaki.hentoid.activities.PrefsActivity;
import me.devsaki.hentoid.activities.bundles.PrefsBundle;
import me.devsaki.hentoid.util.Preferences;

public final class ReaderPrefsDialogFragment extends DialogFragment {

    private static final String RENDERING_MODE = "render_mode";
    private static final String BROWSE_MODE = "browse_mode";
    private static final String DISPLAY_MODE = "display_mode";

    private Parent parent;
    private int renderingMode;
    private int browseMode;
    private int displayMode;


    public static void invoke(Fragment parent, Map<String, String> bookPrefs) {
        ReaderPrefsDialogFragment fragment = new ReaderPrefsDialogFragment();

        Bundle args = new Bundle();
        if (bookPrefs != null) {
            if (bookPrefs.containsKey(VIEWER_RENDERING))
                args.putInt(RENDERING_MODE, Preferences.isContentSmoothRendering(bookPrefs) ? 1 : 0);
            if (bookPrefs.containsKey(VIEWER_BROWSE_MODE))
                args.putInt(BROWSE_MODE, Preferences.getContentBrowseMode(bookPrefs));
            if (bookPrefs.containsKey(VIEWER_IMAGE_DISPLAY))
                args.putInt(DISPLAY_MODE, Preferences.getContentDisplayMode(bookPrefs));
        }
        fragment.setArguments(args);

        fragment.show(parent.getChildFragmentManager(), null);
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        if (null == getArguments()) throw new IllegalArgumentException("No arguments found");
        renderingMode = getArguments().getInt(RENDERING_MODE, -1);
        browseMode = getArguments().getInt(BROWSE_MODE, -1);
        displayMode = getArguments().getInt(DISPLAY_MODE, -1);
        parent = (Parent) getParentFragment();
    }

    @Override
    public void onDestroy() {
        parent = null;
        super.onDestroy();
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedState) {
        return inflater.inflate(R.layout.dialog_reader_book_prefs, container, false);
    }

    @Override
    public void onViewCreated(@NonNull View rootView, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(rootView, savedInstanceState);
        Resources res = rootView.getContext().getResources();

        // == Dropdown lists

        String[] browseModes = getResources().getStringArray(R.array.pref_viewer_browse_mode_entries);
        List<String> browseItems = new ArrayList<>();
        // App pref
        browseItems.add(res.getString(R.string.use_app_prefs, browseModes[Preferences.getViewerBrowseMode()]));
        // Available prefs
        browseItems.addAll(Arrays.asList(browseModes));

        PowerSpinnerView browseSpin = rootView.findViewById(R.id.book_prefs_browse_spin);
        browseSpin.setIsFocusable(true);
        browseSpin.setLifecycleOwner(this);
        browseSpin.setItems(browseItems);
        browseSpin.selectItemByIndex(browseMode + 1);


        String[] renderingModes = getResources().getStringArray(R.array.pref_viewer_rendering_entries);
        List<String> renderingItems = new ArrayList<>();
        // App pref
        renderingItems.add(res.getString(R.string.use_app_prefs, renderingModes[Preferences.isViewerSmoothRendering() ? 1 : 0].replace(" (" + getString(R.string._default) + ")", "")));
        // Available prefs
        for (int i = 0; i < renderingModes.length; i++) {
            // No smooth mode for Android 5
            if (1 == i && Build.VERSION.SDK_INT < Build.VERSION_CODES.M) continue;
            renderingItems.add(renderingModes[i].replace(" (" + getString(R.string._default) + ")", ""));
        }

        PowerSpinnerView renderSpin = rootView.findViewById(R.id.book_prefs_rendering_spin);
        renderSpin.setIsFocusable(true);
        renderSpin.setLifecycleOwner(this);
        renderSpin.setItems(renderingItems);
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.M) {
            renderSpin.selectItemByIndex(0);
            renderSpin.setEnabled(false);
        } else renderSpin.selectItemByIndex(renderingMode + 1);


        String[] displayModes = getResources().getStringArray(R.array.pref_viewer_display_mode_entries);
        List<String> displayItems = new ArrayList<>();
        // App pref
        displayItems.add(res.getString(R.string.use_app_prefs, displayModes[Preferences.getViewerDisplayMode()]));
        // Available prefs
        for (String mode : displayModes) {
            displayItems.add(mode.replace(" (" + getString(R.string._default) + ")", ""));
        }

        PowerSpinnerView displaySpin = rootView.findViewById(R.id.book_prefs_display_spin);
        displaySpin.setIsFocusable(true);
        displaySpin.setLifecycleOwner(this);
        displaySpin.setItems(displayItems);
        displaySpin.selectItemByIndex(displayMode + 1);


        // == Bottom buttons

        View appSettingsBtn = rootView.findViewById(R.id.book_prefs_app_prefs_btn);
        appSettingsBtn.setOnClickListener(v -> {
            Intent intent = new Intent(requireActivity(), PrefsActivity.class);

            PrefsBundle prefsBundle = new PrefsBundle();
            prefsBundle.setViewerPrefs(true);
            intent.putExtras(prefsBundle.getBundle());

            requireContext().startActivity(intent);
        });

        View okBtn = rootView.findViewById(R.id.book_prefs_ok_btn);
        okBtn.setOnClickListener(v -> {
            Map<String, String> newPrefs = new HashMap<>();
            if (renderSpin.getSelectedIndex() > 0)
                newPrefs.put(VIEWER_RENDERING, (renderSpin.getSelectedIndex() - 1) + "");
            if (browseSpin.getSelectedIndex() > 0)
                newPrefs.put(VIEWER_BROWSE_MODE, (browseSpin.getSelectedIndex() - 1) + "");
            if (displaySpin.getSelectedIndex() > 0)
                newPrefs.put(VIEWER_IMAGE_DISPLAY, (displaySpin.getSelectedIndex() - 1) + "");
            parent.onBookPreferenceChanged(newPrefs);
            dismiss();
        });
    }

    public interface Parent {
        void onBookPreferenceChanged(Map<String, String> newPrefs);
    }
}

File: app/src/main/java/me/devsaki/hentoid/fragments/reader/ReaderPrefsDialogFragment.kt
package me.devsaki.hentoid.fragments.reader

import android.content.Intent
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.core.view.isVisible
import androidx.fragment.app.Fragment
import me.devsaki.hentoid.R
import me.devsaki.hentoid.activities.bundles.PrefsBundle
import me.devsaki.hentoid.activities.prefs.PreferencesActivity
import me.devsaki.hentoid.databinding.DialogReaderBookPrefsBinding
import me.devsaki.hentoid.enums.Site
import me.devsaki.hentoid.fragments.BaseDialogFragment
import me.devsaki.hentoid.util.Settings
import me.devsaki.hentoid.util.Settings.Key.VIEWER_BROWSE_MODE
import me.devsaki.hentoid.util.Settings.Key.VIEWER_IMAGE_DISPLAY
import me.devsaki.hentoid.util.Settings.Key.VIEWER_RENDERING
import me.devsaki.hentoid.util.Settings.Value.VIEWER_BROWSE_TTB
import kotlin.math.min

class ReaderPrefsDialogFragment : BaseDialogFragment<ReaderPrefsDialogFragment.Parent>() {
    companion object {

        const val RENDERING_MODE = "render_mode"
        const val BROWSE_MODE = "browse_mode"
        const val TWOPAGES_MODE = "twopages_mode"
        const val DISPLAY_MODE = "display_mode"
        const val SITE = "site"

        fun invoke(parent: Fragment, site: Site, bookPrefs: Map<String, String>) {
            val args = Bundle()
            if (bookPrefs.containsKey(VIEWER_RENDERING)) args.putInt(
                RENDERING_MODE,
                if (Settings.isContentSmoothRendering(bookPrefs)) 1 else 0
            )
            if (bookPrefs.containsKey(VIEWER_BROWSE_MODE)) args.putInt(
                BROWSE_MODE,
                Settings.getContentBrowseMode(site, bookPrefs)
            )
            if (bookPrefs.containsKey(Settings.Key.READER_TWOPAGES)) args.putBoolean(
                TWOPAGES_MODE,
                Settings.getContent2PagesMode(bookPrefs)
            )
            if (bookPrefs.containsKey(VIEWER_IMAGE_DISPLAY)) args.putInt(
                DISPLAY_MODE,
                Settings.getContentDisplayMode(site, bookPrefs)
            )
            args.putInt(SITE, site.code)
            invoke(parent, ReaderPrefsDialogFragment(), args)
        }
    }

    // UI
    private var binding: DialogReaderBookPrefsBinding? = null

    // === VARIABLES
    private var bookRenderingMode = 0
    private var bookBrowseMode = 0
    private var bookDisplayMode = 0
    private var bookTwoPagesMode = false
    private var hasSiteBrowseMode = false
    private var siteBrowseMode = 0

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        requireNotNull(arguments) { "No arguments found" }
        bookRenderingMode = requireArguments().getInt(RENDERING_MODE, -1)
        bookBrowseMode = requireArguments().getInt(BROWSE_MODE, -1)
        bookDisplayMode = requireArguments().getInt(DISPLAY_MODE, -1)
        bookTwoPagesMode = requireArguments().getBoolean(TWOPAGES_MODE, false)
        val site = Site.searchByCode(requireArguments().getInt(SITE, Site.ANY.code).toLong())
        siteBrowseMode = Settings.getReaderBrowseMode(site)
        hasSiteBrowseMode = siteBrowseMode != Settings.appReaderBrowseMode
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedState: Bundle?
    ): View? {
        binding = DialogReaderBookPrefsBinding.inflate(inflater, container, false)
        return binding?.root
    }

    override fun onDestroyView() {
        binding = null
        super.onDestroyView()
    }

    override fun onViewCreated(rootView: View, savedInstanceState: Bundle?) {
        super.onViewCreated(rootView, savedInstanceState)

        val res = rootView.context.resources

        // == Dropdown lists
        val browseModes = resources.getStringArray(R.array.pref_viewer_browse_mode_entries)
        val browseItems = ArrayList<String>()

        // App pref
        browseItems.add(
            res.getString(R.string.use_app_prefs, browseModes[Settings.appReaderBrowseMode])
        )
        // Site pref, if different
        if (hasSiteBrowseMode)
            browseItems.add(res.getString(R.string.use_source_prefs, browseModes[siteBrowseMode]))
        // Available prefs
        browseItems.addAll(listOf(*browseModes))

        binding?.apply {
            browsePicker.entries = browseItems
            browsePicker.index = min(bookBrowseMode + 1, browsePicker.entries.size - 1)
            browsePicker.setOnIndexChangeListener { refreshValues() }
            twoPagesSwitch.isChecked = bookTwoPagesMode
        }

        val renderingModes = resources.getStringArray(R.array.pref_viewer_rendering_entries)
        val renderingItems: MutableList<String> = ArrayList()
        // App pref
        renderingItems.add(
            res.getString(
                R.string.use_app_prefs,
                renderingModes[if (Settings.isReaderSmoothRendering()) 1 else 0].replace(
                    " (" + getString(R.string._default) + ")", ""
                )
            )
        )
        // Available prefs
        for (i in renderingModes.indices) {
            renderingItems.add(
                renderingModes[i].replace(
                    " (" + getString(R.string._default) + ")",
                    ""
                )
            )
        }
        binding?.apply {
            renderingPicker.entries = renderingItems
            renderingPicker.index = bookRenderingMode + 1
        }

        val displayModes = resources.getStringArray(R.array.pref_viewer_display_mode_entries)
        val displayItems: MutableList<String> = ArrayList()
        // App pref
        displayItems.add(
            res.getString(
                R.string.use_app_prefs,
                displayModes[Settings.readerDisplayMode]
            )
        )
        // Available prefs
        for (mode in displayModes) {
            displayItems.add(mode.replace(" (" + getString(R.string._default) + ")", ""))
        }
        binding?.apply {
            displayPicker.entries = displayItems
            displayPicker.index = bookDisplayMode + 1
        }

        // == Bottom buttons
        binding?.appPrefsBtn?.setOnClickListener {
            val intent = Intent(requireActivity(), PreferencesActivity::class.java)
            val prefsBundle = PrefsBundle()
            prefsBundle.isViewerPrefs = true
            intent.putExtras(prefsBundle.bundle)
            requireContext().startActivity(intent)
        }

        binding?.actionButton?.setOnClickListener {
            val newPrefs: MutableMap<String, String> = HashMap()
            binding?.apply {
                if (renderingPicker.index > 0) newPrefs[VIEWER_RENDERING] =
                    (renderingPicker.index - 1).toString()
                if (browsePicker.index > 0) newPrefs[VIEWER_BROWSE_MODE] =
                    (browsePicker.index - 1).toString()
                newPrefs[Settings.Key.READER_TWOPAGES] = twoPagesSwitch.isChecked.toString()
                if (displayPicker.index > 0) newPrefs[VIEWER_IMAGE_DISPLAY] =
                    (displayPicker.index - 1).toString()
            }
            parent?.onBookPreferenceChanged(newPrefs)
            dismiss()
        }
    }

    private fun refreshValues() {
        binding?.apply {
            twoPagesSwitch.isVisible = (VIEWER_BROWSE_TTB != browsePicker.index - 1)
            if (0 == browsePicker.index && Settings.appReaderBrowseMode == VIEWER_BROWSE_TTB)
                twoPagesSwitch.isVisible = false
            if (!twoPagesSwitch.isVisible) twoPagesSwitch.isChecked = false
        }
    }

    interface Parent {
        fun onBookPreferenceChanged(newPrefs: Map<String, String>)
    }
}